parla/cuda.py

```diff
@@ -50,13 +52,28 @@ class _GPUMemory(Memory):
     def np(self):
         return _DeviceCUPy(self.device)
 
+    def asarray_sync(self, src):
+      dst = cupy.ndarray(src.shape)
+      dst.data.copy_from_device_async(src.data, src.nbytes,
+                                      cupy.cuda.get_current_stream())
+      return dst
+
     def __call__(self, target):
-        with self.device._device_context():
-            if cupy.cuda.Device() != getattr(target, "device", None):
-                logger.debug("Moving data: %r => %r", getattr(target, "device", None), cupy.cuda.Device())
-                return cupy.asarray(target)
-            else:
-                return target
+        # TODO Several threads could share the same device object.
+        #      It causes data race and CUDA context is incorrectly set.
+        #      For now, this remove assumes that one device is always
+        #      assigned to one task.
+        #with self.device._device_context():
```

Comment:
Just pushed an alternate fix for the context manager issue to the other PR. LMK what you think. Otherwise, LGTM.

Reply:
Fixed!

Reply:
The code in this PR is actually wrong because this function can be called with a different device active to copy data to a remote device instead of to the current device. Let's go with Ian's solution if at all possible.

Reply:
I see in https://github.com/ut-parla/Parla.py/pull/20#issuecomment-810648174 that Ian's fix is not enough. So do whatever hack you need but make sure to document that this is breaking the semantics of the function in a big way, so we can go back and fix it.